src/metaprob/autodiff.cljc

```diff
@@ -0,0 +1,247 @@
+(ns metaprob.autodiff
+  (:refer-clojure :exclude [+ * - / == < > <= >= zero? number? pos? neg?]))
+
+;; Dual numbers represent computations dependent on a value with respect to which we are differentiating.
+;; derivative contains the current derivative; value contains the current value; and tag is a special tag
+;; that helps distinguish between multiple "levels" of nested differentiations.
+(defn make-dual-number [tag value derivative]
+  {:tag tag :value value :derivative derivative})
+
+(declare * + - /)
+
+(def non-dual? clojure.core/number?)
+
+;; Take a unary operation f on numbers
+;; and lift it to work with dual numbers.
+;; Note that df-dx should itself be a lifted
+;; function for computing the derivative of f
+;; at a point.
+(defn lift-real->real
+  [f df-dx]
+  (fn new-f [x]
+    (if (non-dual? x)
+      (f x)
+      (make-dual-number (:tag x)
+                        (new-f (:value x))
```

Comment:
Please help me understand - by recursion, what data type (and how deeply nested) do we expect `(new-f (:value x))` to evaluate to?

Reply:
Applying `new-f` to `x` returns the same datatype as `x`. Applying `new-f` to `(:value x)` returns something that is nested one layer _less_ deep than `x` is. In one typical case, `x` might be a  dual number,  `(:value x)` might be a non-dual number, and so `(f (:value x))` would also be a non-dual number.

Reply:
Is it possible that `(:value x)` is itself a dual number?  I.e. is there tree structure?

Reply:
Yep! This could happen if we are taking higher-order derivatives. The purpose of the `:tag` is to distinguish between the different differentiations that might all be happening simultaneously.

Reply:
Cool.  Where does the construction of the new tag happen?  AFAICS, lifting twice results in using the same tag as lifting once.

Reply:
The "lift" operation creates a new version of an operator that works well with the AD framework; it does not create a "differentiated" version of the operator. Neither lifting the operator nor calling the lifted operator "starts" a new differentiation, and as such, tags are not created when performing either of these operations.

Tags are created when the user calls `forward-mode` to actually perform differentiation, or when they use the helper functions `diff` or `gradient`.